### DISCUSSION TITLE
PictView: Support WEBP and SVG Files

### DISCUSSION THREAD

**Discussion by melloware** [OpenSalamander/salamander#36] [Ideas]
WebP is a [raster graphics](https://en.wikipedia.org/wiki/Raster_graphics) [file format](https://en.wikipedia.org/wiki/Image_file_format) developed by [Google](https://en.wikipedia.org/wiki/Google) intended as a replacement for [JPEG](https://en.wikipedia.org/wiki/JPEG), [PNG](https://en.wikipedia.org/wiki/Portable_Network_Graphics), and [GIF](https://en.wikipedia.org/wiki/GIF) file formats. It supports both [lossy](https://en.wikipedia.org/wiki/Lossy_compression) and [lossless](https://en.wikipedia.org/wiki/Lossless_compression) compression,[[8]](https://en.wikipedia.org/wiki/WebP#cite_note-8) as well as [animation](https://en.wikipedia.org/wiki/Computer_animation) and [alpha transparency](https://en.wikipedia.org/wiki/Alpha_compositing).

Google announced the WebP format in September 2010, and released the first stable version of its supporting [library](https://en.wikipedia.org/wiki/Library_(computing)) in April 2018.

1. Add `*.webp` to the default PictViewer list

2. Support WEBP in the Pictview Plugin

3. Add `*svg` to the default PictViewer list

4. Support SVG in the Pictview Plugin

**Comment by lejcik** [comment]
This is something what was missing in PictView plugin - support for newer image file types. I added an experimental support for `.heif` and `.webp` image files. The plugin would need a bit more tuning, but it's quite useful yet. Whoever is interested, you can find it in my fork in this branch:

https://github.com/lejcik/salamander/tree/plugin-pictview-support-for-heif-and-webp

The code structure of the plugin is very complex. Firstly I tried out to modify it to add support for other image file types, but when I changed something, another issue appeared on the other end. This is kind of _"please don't touch this code!"_ legacy and monolithic code... This is not a complaint, my codes looked similar 20 years ago 😀

Plugin uses pict-view library, and I found out that plugin can paste an image from clipboard. Then I found a way how the library can read a bitmap from memory. So, if I open an arbitrary image file and convert it to a bitmap, the pict-view library will accept it. My implementation replaces API of the library. With minimal code modification I was able to add new image loaders.

This approach has some limitations - only raster images are supported (no vector images such as svg), only primary image can be load (no animated images), no save-as in the new image filetypes (only existing ones), etc. But I think for the most cases this is ok.

Here are a few issues I have to resolve:

- load `libheif.dll` and `libwebp.dll` dynamically (currently they have to be copied to the directory where `salamand.exe` is),
- libheif: not sure how to cancel `heif_decode_image()`, image decoding may take a few seconds and cannot be interrupted yet,
- can't show image loading progress (image has to be loaded and decoded, and only then a bitmap can be shown),
- read thumbnail, if it's present in the file,
- read EXIF, if it's present in the file,
- printing on x64 causes crash of the background pict-view process,
- consider adding support for other raster image file types.

Any feedback is welcome. Enjoy 😉

**Reply by melloware** [reply]
Can you submit a PR?

**Reply by lejcik** [reply]
It's still not ready for pull request, but ok, I created one, and I will keep it as a draft: https://github.com/OpenSalamander/salamander/pull/61

**Reply by melloware** [reply]
OK i found this awesome single file C library for SVG: https://github.com/memononen/nanosvg

**Comment by KRtekTM** [comment]
https://github.com/KRtkovo-eu-AI/salamander/tree/plugins_darkmode/src/plugins/webview2renderviewer

# WebView2 Render Viewer .NET Supported File Types

| Category | Extensions |
| --- | --- |
| HTML and web archives | `.html`, `.htm`, `.xhtml`, `.mhtml`, `.mht` |
| Markdown (rendered to HTML) | `.md`, `.markdown`, `.mdown`, `.mkd`, `.mdx` |
| SVG vector graphics | `.svg`, `.svgz` |
| Modern image formats | `.webp`, `.avif`, `.apng` |
| Raster image formats | `.png`, `.jpg`, `.jpeg`, `.jfif`, `.gif`, `.bmp`, `.ico`, `.tif`, `.tiff` |
| Portable Document Format | `.pdf` |

All matches are case-insensitive. Files larger than 32 MB are rejected to match the plug-in limits.

**Reply by melloware** [reply]
@KRtekTM  this is amazing change.  Do you have a build I can test this with?

**Reply by melloware** [reply]
@KRtekTM i think this deserves its own PR here this is a great addition

**Reply by KRtekTM** [reply]
> @KRtekTM this is amazing change. Do you have a build I can test this with?

Yes, here: https://github.com/KRtkovo-eu-AI/salamander/releases/tag/5.0-samandarin-0.1

> @KRtekTM i think this deserves its own PR here this is a great addition

Not the production quality yet.

**Reply by melloware** [reply]
Tested and SVG and WebP work great.

**Comment by KRtekTM** [comment]
So there is also PictView with backend switched to WIC (Windows Imaging Component): https://github.com/user-attachments/files/22840524/plugin_5.0_pictview_2.2_80c1a42.zip

With [WebP codec](https://apps.microsoft.com/detail/9pg2dk419drg) installed, it's working out of the box.
https://learn.microsoft.com/en-us/windows/win32/wic/native-wic-codecs

**Comment by 0xeb** [comment]
> WebP is a [raster graphics](https://en.wikipedia.org/wiki/Raster_graphics) [file format](https://en.wikipedia.org/wiki/Image_file_format) developed by [Google](https://en.wikipedia.org/wiki/Google) intended as a replacement for [JPEG](https://en.wikipedia.org/wiki/JPEG), [PNG](https://en.wikipedia.org/wiki/Portable_Network_Graphics), and [GIF](https://en.wikipedia.org/wiki/GIF) file formats. It supports both [lossy](https://en.wikipedia.org/wiki/Lossy_compression) and [lossless](https://en.wikipedia.org/wiki/Lossless_compression) compression,[[8]](https://en.wikipedia.org/wiki/WebP#cite_note-8) as well as [animation](https://en.wikipedia.org/wiki/Computer_animation) and [alpha transparency](https://en.wikipedia.org/wiki/Alpha_compositing).
> 
> Google announced the WebP format in September 2010, and released the first stable version of its supporting [library](https://en.wikipedia.org/wiki/Library_(computing)) in April 2018.
> 
> 1. Add `*.webp` to the default PictViewer list
> 2. Support WEBP in the Pictview Plugin
> 3. Add `*svg` to the default PictViewer list
> 4. Support SVG in the Pictview Plugin

https://github.com/0xeb/salamander/issues/27#issuecomment-3939810785
